### PR TITLE
Consolidation of LevelDB database handling

### PR DESCRIPTION
**Pull request by dexX7**
**LevelDB consolidation**

This PR introduces a new base class `CDBBase` for LevelDB based storage, which is used by `CMPSTOList`, `CMPTradeList`, `CMPTxList` and `CMPSPInfo`.

`CDBBase` allows to open, or create, LevelDB databases via `Open()`, which is restricted to subclasses, to prevent the accidental opening of databases more than once. The methods `Clear()` and `Close()` were added to explicitly clear the database, or to close the database explicitly. It is not required to call `Close()` during shutdown, and the databases are properly closed, once the object is destroyed.

`NewIterator()` can be used as shortcut, to create a new LevelDB iterator, which uses appropriate options to iterate through entries (i.e. `fill_cache = false`). Altogether, this consolidates LevelDB database handling, and should make it more easy to work with data storage.

**Resetting the state properly**

Up until now, only the property DB was cleared, when the state was resetted, for example during a reorganization, which can result in faulty entries.

**Avoid using boost::format**

`boost::format` was replaced by `tinyformat`, which is heavily used in the whole project already, and which should be slightly faster. Some applications, especially in `mastercore_sp`, were optimized in this context. As bonus, this also resolves some memory access violations related to `MPPersistencePath`, which were appearingly not harmful, but detected by valgrind.

**Initialization behavior**

At the very moment, the initialization behavior is unchanged, and while `fReindex` should wipe the databases during initialization, it never actually did. However, when wiping the databases during a reindex, an issue is triggered, which is related to the block handlers. This issue should be addresses separately, but may ideally be discussed here.

**Tests and mainnet consensus comparison passed**

This PR passes all tests and is in consensus on mainnet (locally tested). I also tested different scenarios, where I stopped the initial parsing and started from the beginning.

### CODE REVIEW

**Comment by dexX7** [comment]
The `fReindex` issue:

When the databases are wiped during a reindex, then the resulting state differs significantly from the actual state. It's not entierly clear to me why, and when then restarting again with `-startclean`, everything is fine.

This appears to be related to the block disconnect handlers, which are skipped during a reindex, and with the following change, the correct state is maintained, even when the databases are wiped during a reindex (as they should!):
```diff
@@ -1968,5 +1968,3 @@ bool static DisconnectTip(CValidationState &state) {
     LogPrint("handler", "Omni Core handler: block disconnect begin [height: %d, reindex: %d]\n", GetHeight(), (int)fReindex);
-    if (!fReindex) {
-        mastercore_handler_disc_begin(GetHeight(), pindexDelete);
-    }
+    mastercore_handler_disc_begin(GetHeight(), pindexDelete);
 
@@ -1980,5 +1978,3 @@ bool static DisconnectTip(CValidationState &state) {
     LogPrint("handler", "Omni Core handler: block disconnect end [height: %d, reindex: %d]\n", GetHeight(), (int)fReindex);
-    if (!fReindex) {
-        mastercore_handler_disc_end(GetHeight(), pindexDelete);
-    }
+    mastercore_handler_disc_end(GetHeight(), pindexDelete);
```

I did not include this change in this PR, and mirrored the original initialization behavior of Omni Core,  because I have not really an understanding of the actual issue, and it seemed to risky to mess with it. Appearingly some of the reorganization logic kicks in during a reindex, as there are several `isMPinBlockRange` and `deleteAboveBlock` calls.

In the context of block handlers, there is at least one other issue, fixed via https://github.com/zathras-crypto/mastercore/commit/b802850bdf68b8eba51405eee5ff9c0324d627d1 in the audit branch, and https://github.com/zathras-crypto/mastercore/commit/4ef9c6de3891b81c97d5e0c5700f0fd15ddd5836 appears to tackle some of the reorganization logic. I'd apprechiate some insight and details about the later, because it's not 100 % clear to me, what issue it solves.

Further discussion about this, which motivated this PR, starts here: https://github.com/zathras-crypto/mastercore/issues/66#issuecomment-95977576

**Comment by zathras-crypto** [comment]
Mate, you've been killing it the last couple of days, just wanted to say a big thank you - really, really appreciate it :)

Reviewing this now...

**Comment by zathras-crypto** [comment]
> I'd apprechiate some insight and details 

Sure :)  So for zathras-crypto@4ef9c6d - basically when we reorg we roll back the state (assuming we can retrieve said state from persistence), we are rolling back to a specific block height.  We thus make calls out to the various LevelDB databases to delete any entries in blocks above this block height.

What that commit does is fix the height that is being passed into the deletion functions - I can probably explain it best with an example:

0) Chain is at Block 201
1) Block 201 is orphaned
2) State is rolled back to Block 200.
3) Calls are made to LevelDB to delete any info about transactions and trades after Block 200.

The issue was that the same parameter (```pBlockIndex->nHeight```) was passed into each deletion function, but the deletion functions handle inclusiveness differently.  ```isMPinBlockRange``` is looking for a range of blocks and *includes* the start and finish blocks supplied as parameters.  ```deleteAboveBlock``` deletes stuff for blocks *above* the supplied parameter (so does not *include* the supplied parameter).

In a nutshell:
* We pass 201 and say 205 into ```isMPinBlockRange``` and it will delete all entries with block heights in the range (ie 201,202,203,204,205)
* We pass 201 into ```deleteAboveBlock``` and it will delete all entries with block heights *above* 201 (ie 202,203,204,205,206 and so on)

Before Commit:
```
isMPinBlockRange(201, 201) - deletes stuff with block height of 201
deleteAboveBlock(201) - deletes stuff with block height *over* 201 so does not include block 201
deleteAboveBlock(201)
```
After Commit:
```
isMPinBlockRange(201, 201) - deletes stuff with block height of 201
deleteAboveBlock(200) - deletes stuff with block height *over* 200 so now includes block 201
deleteAboveBlock(200)
```

**Comment by zathras-crypto** [comment]
PR looks good.  Thanks again.

**Comment by zathras-crypto** [comment]
Also, will look into the reindex stuff shortly.

**Comment by zathras-crypto** [comment]
> This appears to be related to the block disconnect handlers, which are skipped during a reindex

@m21 can you shed any light on this?

@dexX7 skipping block disconnect handlers during a reindex does not make any sense to me.  A reindex is basically ditching the index and chainstate and reprocessing the chain - just from disk instead of from peers - orphaned blocks persist in blk?????.dat files and so would be handled during a reindex just the same as if they came in from peers.

Eg, if your blk?????.dat files contain:
Block 1
Block 2
Block 3 (orphaned)
Block 3
Block 4

Every time you reindex you will see the first block 3 get added to the chain, then see it get orphaned.  Reindexing (AFAIK) is just processing the chain from disk instead of peers.

Now with all that said, I don't know how 0.10 out of order block storage affected that process.

TL:DR; it makes perfect sense to remove those bypasses on the handlers.  What interests me is how they got there in the first place and the reasoning for them.

**Comment by zathras-crypto** [comment]
Found another issue - ```clear_all_state``` still isn't quite complete bud, could you please clear the MetaDEx maps too under ```// Memory based storage```?

**Comment by dexX7** [comment]
> What that commit does is fix the height that is being passed into the deletion functions ...

So, what worries me is basically: it changes the logic significantly, and enabling the block disconnect handlers durng a reindex is even more significant, and this tells me two things:

- we were probably lucky, that we haven't run into problems yet
- our tests are not sufficient, given that this slipped through

> Found another issue ...

Very good catch! Whlie testing and playing around with the meta DEx in general, I also noticed, even if a trade is removed, the other maps (priceMap and propertyMap) are not updated, but it looks like this isn't an issue.

I'll pickup the clearing of the `metadex` object, but I suggest to tackle the handlers in a different PR.

@zathras-crypto: in general, there are more than one useful commits in your branches. Could you please extract those, and push them in a seperate PR, which we can merge directly (independend from the audit branch)? For example: the reorganization fix?

**Comment by zathras-crypto** [comment]
> So, what worries me is basically: it changes the logic significantly

Actually not so much (re ```deleteAboveBlock``` changes) - basically without the fix we may miss deleting transactions from the STO receipts DB and the completed trades DB, but these would simply be overwritten when those transactions were reprocessed.  Neither of these databases is used for any logic, only for reporting on the RPC and UI layers.  Still it meant we could give out bad data if such a query was made after transaction(s) had been orphaned but before they were re-included in another block, so I wanted to fix it :)

> enabling the block disconnect handlers durng a reindex is even more significant

Yeah, this will 100% affect the logic during a reindex, but from what I can see the current logic during a reindex is broken?  I think ensuring we provide a correct state whether reindexing or not has to be the primary driver.  As it stands I'm also surprised that we haven't had trouble with this before - I would posit most integrators have simply spun up a fresh instance and allowed it to download the chain itself rather than applying OmniCore to an existing un-indexed chain - still we should fix though.

> Could you please extract those, and push them in a seperate PR

Yeah will do mate :)

**Comment by dexX7** [comment]
> Neither of these databases is used for any logic, only for reporting on the RPC and UI layers

Let me play devil's advocate: from an user's point of view, there is probably no difference.

> these would simply be overwritten when those transactions were reprocessed

Given the proper clearing of state, I assume the overwrite protection should be removed or refined? It might be thinkable to check, whether the key to insert exists, and if so, whether the value to insert equals the existing value. If this is not the case, something is probably very wrong, but I can't think of a situation, where this could happen in practise.

> Yeah, this will 100% affect the logic during a reindex, but from what I can see the current logic during a reindex is broken?

Right now (and before) the databases were not wiped during a reindex, and if we do so, we must ensure all data is properly reinserted, and available, to begin with. Is this certainly the case? For example, I'm thinking about [prune_state_files()](https://github.com/OmniLayer/omnicore/blob/91acb51fb8af59bbd813e909ffbbe8e3aa99fff2/src/mastercore.cpp#L2263) in this context.

**Comment by zathras-crypto** [comment]
> I assume the overwrite protection should be removed or refined

I'd actually been thinking of moving levelDB overwrite detection to the auditor, and expanding it to cover all databases (currently it only covers ```txlistdb```) rather than removing it - I believe it still serves a purpose in identifying when things are going wrong.

> Right now (and before) the databases were not wiped during a reindex, and if we do so, we must ensure all data is properly reinserted, and available, to begin with

Bit unsure what you mean here mate, a reindex by it's nature means a fresh parse - there is no way to start a reindex from a certain block - thus there is no data to reinsert, we should be starting with a clean slate no?

> Let me play devil's advocate: from an user's point of view, there is probably no difference.

Hehe fair point, I was just noting that errors or duplicates in those databases cannot result in an incorrect movement of tokens/incorrect balances (cannot affect the logic) - of course I do agree that it's a problem to give incorrect trade history or STO receipts I just meant it couldn't corrupt the logic and put tokens in the wrong place etc.

> Merged

Nice mate thanks again, great work here :)

**Comment by dexX7** [comment]
> I'd actually been thinking of moving levelDB overwrite detection to the auditor

This is an excellent idea!

> Bit unsure what you mean here mate, a reindex by it's nature means a fresh parse ...

When reindexing, the `MP_` databases [were never](https://github.com/zathras-crypto/mastercore/issues/66#issuecomment-96123963), and are [currently neither](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore.cpp#L2409-L2419) wiped during the initialization, and I was solely referring to our databases, and not the Bitcoin ones. I agree that they should be wiped here, but this yielded a very wrong state, likely due to the block disconnect handlers, so I basically didn't touch (or rather change) the init-wipe-during-reindex-behavior in this PR.

**Comment by zathras-crypto** [comment]
Yep, I get what you're saying - basically what I'm getting at is:

* The block handler bypasses should not be there (unless I've severely misunderstood the reindexing process)
* The state should be completely wiped prior to a reindex

Both I consider bugs.

> but this yielded a very wrong state

This is what I don't get - can I just clarify mate?  If you *both* remove the block handlers, *and* wipe the state on a reindex, then you end up with a bad state?

**Comment by dexX7** [comment]
> If you both remove the block handlers, and wipe the state on a reindex, then you end up with a bad state?

Ahh. :) Yes, when the databases are wiped *and* the disconnect handlers used during a reindex, then the state appears to be fine - at least this is what I observed in regtest mode and on testnet.

**Comment by zathras-crypto** [comment]
Ah cool OK :) thanks - yeah I don't think it can be one or the other - the two changes need to go in simultaneously.  

I assume the bypassed block handlers are a way to sort of resume from the persistent state on completion of the reindex, but I'm just guessing and it seems kind of 'hacky' - I'd much prefer to pull us back to a clean state and parse during a reindex just as if the chain was being downloaded which is what removing the disconnect handler bypasses & wiping the state should do.
